File: crates/scematica-sentience/src/cognitive_loop.rs
//! §24  Recursive Cognitive Loop
//!
//! ```text
//! Ω_t → Perception → Integration → Reasoning → Prediction → EthicalEvaluation
//!      → Decision → Action → Observation → Error → Learning → Memory
//!      → SelfModel → Ω_{t+1}
//! ```

use serde::{Deserialize, Serialize};
use crate::{
    cognitive_state::CognitiveState,
    error_correction::ErrorCorrector,
    learning::Learner,
    sentience::SentienceIndex,
    master_equation::MasterEquation,
    types::{Bounded, Observation},
};

/// Output of one full cognitive cycle.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CycleOutput {
    pub timestep: u64,
    pub sentience: SentienceIndex,
    pub psi: Bounded,
    pub learning_delta: f64,
    pub error: f64,
    pub reassessment_triggered: bool,
}

/// Drives the recursive Ω_{t+1} = F(Ω_t, ...) loop.
pub struct CognitiveLoop {
    pub state: CognitiveState,
    pub learner: Learner,
    pub error_corrector: ErrorCorrector,
}

impl CognitiveLoop {
    pub fn new(state: CognitiveState) -> Self {
        Self {
            state,
            learner: Learner::default(),
            error_corrector: ErrorCorrector::default(),
        }
    }

    /// Run one full cognitive cycle given an observation and predicted value.
    pub fn step(&mut self, observation: Observation, predicted: f64, feedback: f64) -> CycleOutput {
        // 1. Learning
        let update = self.learner.update(&observation, predicted);

        // 2. Error correction
        let correction = self.error_corrector.step(
            observation.value,
            predicted,
            observation.confidence,
        );

        // 3. Master equation
        let (sentience, psi) = MasterEquation::compute(
            &self.state.rationality,
            &self.state.logic,
            &self.state.ethics,
            &self.state.perception,
            &self.state.agency.inputs,
            &crate::meta_cognition::MetaCognitionInputs::default(),
            self.state.knowledge_density,
            Bounded::new(feedback),
        );

        // 4. Update state
        self.state.sentience = sentience;
        self.state.last_observation = Some(observation);
        self.state.tick();

        CycleOutput {
            timestep: self.state.timestep,
            sentience,
            psi: psi.psi,
            learning_delta: update.delta,
            error: correction.raw_error,
            reassessment_triggered: update.triggered_reassessment,
        }
    }
}

File: alchem-link/src/sentience/cognitive_loop.py
"""§24 Recursive cognitive loop — one full Ω_{t+1} = F(Ω_t,...) step."""
from __future__ import annotations
from dataclasses import dataclass
from .types import Bounded, Observation
from .cognitive_state import CognitiveState
from .sentience import SentienceIndex
from .master_equation import (
    MasterEquation, DEFAULT_AGENCY_RATIO, DEFAULT_META_RATIO,
)


@dataclass
class CycleOutput:
    timestep: int
    sentience: SentienceIndex
    psi: Bounded
    learning_delta: float
    error: float
    reassessment_triggered: bool


class CognitiveLoop:
    def __init__(self, state: CognitiveState, learning_rate: float = 0.1,
                 reassessment_threshold: int = 3):
        self.state = state
        self.alpha = learning_rate
        self.knowledge = state.knowledge_density.value
        self._streak = 0
        self._thresh = reassessment_threshold

    def step(self, observation: Observation, predicted: float,
             feedback: float = 1.0) -> CycleOutput:
        # Learning
        err = observation.value - predicted
        delta = self.alpha * observation.confidence.value * err
        self.knowledge = max(0.0, min(1.0, self.knowledge + delta))

        # Error streak → reassessment
        if abs(err) > 0.2:
            self._streak += 1
        else:
            self._streak = 0
        reassess = self._streak >= self._thresh

        # Master equation
        sentience, psi = MasterEquation.compute(
            self.state.rationality, self.state.logic,
            self.state.ethics, self.state.perception,
            agency_ratio=DEFAULT_AGENCY_RATIO, meta_ratio=DEFAULT_META_RATIO,
            knowledge_density=self.knowledge,
            feedback=max(0.0, min(1.0, feedback)),
        )

        self.state.sentience = sentience
        self.state.knowledge_density = Bounded(self.knowledge)
        self.state.last_observation = observation
        self.state.tick()

        return CycleOutput(
            timestep=self.state.timestep,
            sentience=sentience,
            psi=psi.psi,
            learning_delta=delta,
            error=abs(err),
            reassessment_triggered=reassess,
        )
